python/iron/placers.py

```diff
@@ -122,11 +233,129 @@ def _get_common_col(self, tiles: list[Tile]) -> int:
         avg_col = round(statistics.mean(cols))
         return avg_col
 
-    def _find_col_match(self, col: int, tiles: list[Tile]) -> Tile:
+    def _find_col_match(self, col: int, tiles: list[Tile], device: Device) -> Tile:
         """
         A utility function that sequentially searches a list of tiles to find one with a matching column.
+        The column is increased until a tile is found in the device, or an error is signaled.
+        """
+        new_col = col
+        while new_col < device.cols:
+            for t in tiles:
+                if t.col == new_col:
+                    return t
+            new_col += 1
+        raise ValueError(
+            f"Failed to find a tile matching column {col}: tried until column {new_col}. Try using a device with more columns."
+        )
+
+    def _update_channels(
+        self,
+        ofe: ObjectFifoEndpoint,
+        tile: Tile,
+        output: bool,
+        num_required_channels: int,
+        channels: dict[Tile, (ObjectFifoEndpoint, int)],
+        tiles: list[Tile],
+        device: Device,
+        is_shim: bool = False,
+    ):
         """
-        for t in tiles:
-            if t.col == col:
-                return t
-        raise ValueError(f"Failed to find a tile matching column {col}")
+        A utility function that updates given channel and tile lists.
+        """
+        if num_required_channels == 0:
+            return
+        if not tile in channels:
+            channels[tile] = [(ofe, num_required_channels)]
+        else:
+            channels[tile].append((ofe, num_required_channels))
+        used_channels = 0
+        for _, c in channels[tile]:
+            used_channels += c
+        max_tile_channels = device.get_num_connections(tile, output)
+
+        # TODO temporary workaround: should be handled in the target model
+        if is_shim:
+            max_tile_channels = 2
```

Comment:
Is handled. At a shim the switch box will be connected to North/South depending on the direction because it is connected to the ShimMux. Please use the ShimMux to determine the number of DMA channels for Shims: 

https://github.com/Xilinx/mlir-aie/blob/ea9b4dfe7ea91f09c5c29c4d51ca74baea2dc4aa/lib/Dialect/AIE/IR/AIETargetModel.cpp#L226-L242

https://github.com/Xilinx/mlir-aie/blob/ea9b4dfe7ea91f09c5c29c4d51ca74baea2dc4aa/lib/Dialect/AIE/IR/AIETargetModel.cpp#L244-L260